Additional Site Data Type Cleanups.

kennethtran93 opened the pull request:
Closes #792 as parent Issue to additional site data types cleanable by hostnames.

This overhauls the localStorage auto cleanup function and generalizes it for additional site data types:
- Cache (Closes #551 and Possibly Closes #283)
  - Firefox 78+ & Chrome.
- IndexedDB (Closes #171)
  - Firefox 77+ & Chrome.
- Plugin Data (Closes #792, Closes #264 re: Flash Cookies)
  - Firefox 78+ & Chrome.
- Service Workers (Closes #287)
  - Firefox 77+ & Chrome.

*For backwards compatibility, existing expressions will default to keep these new site data types*


Additional items:
- Updated Settings section relating to site data types.
- Returns domains cleaned per site data type to be added to activity log (FUTURE COMMITS/VERSIONS/UPDATE).
- Tests included.
- Fixed some eslint warnings.
- Made Expression Options show 'until restart' localized when listType is GREY.  Should fix some initial user understandings.
- Clean Button dropdown styling update. (alternate result from #793)

New Changes committed as of July 16th:
- New Default Expression Options are now moved to an actual expression entry itself.  See related documentation.
- (Automatic) Site Data Cleanup triggers are now reflected in Activity Log Entries.
- Notifications for Site Data Cleanup - manual and automatic.  This resolves Issue/Request #339.
- Undo button will not show for entries with no cookies to restore.
- Remove individual log entries.
- Migration of settings between 3.4.0 and 3.5.0 - should work if for any reason a downgrade is required, minus the added items.

New Changes committed on July 20th:
- Documentation and FAQ links added to Welcome Page. Resolves #275

New Changes committed on July 23rd:
- Fixed notification regression bug since last commit.
- Added option for showing manual notifications (clean site data for a domain).  Closes #779.
- Reworded some English strings to include other site data other than Cookies as we will have evolved.

Signed-off-by: Kenneth Tran <kennethtran93@users.noreply.github.com>

kennethtran93 committed:
More site data cleanup by hostname.

- Returns domains cleaned per site data type to be added to activity log in the future.
- Tests included.
- Fixed some eslint warnings.
- Made Expression Options show 'until restart' localized when listType is GREY.
- Clean Button dropdown styling update.

Signed-off-by: Kenneth Tran <kennethtran93@users.noreply.github.com>

kennethtran93 committed:
Added Release Notes

Signed-off-by: Kenneth Tran <kennethtran93@users.noreply.github.com>

kennethtran93 committed:
Added Release Notes

Signed-off-by: Kenneth Tran <kennethtran93@users.noreply.github.com>

kennethtran93 committed:
Mention more PR-related resolved.

Signed-off-by: Kenneth Tran <kennethtran93@users.noreply.github.com>

kennethtran93 commented:
For those who want visual changes:

- Updated Settings
![Screenshot_20200707_121650](https://user-images.githubusercontent.com/6724477/86831627-fabd1f80-c04b-11ea-9332-a233d8c5f45f.png)

- Updated Expression Options.
![Screenshot_20200706_164544](https://user-images.githubusercontent.com/6724477/86669546-8d0be780-bfa8-11ea-8cb9-9d1f89cd7c89.png)

- Updated Clean Button Styling
![Screenshot_20200706_164709](https://user-images.githubusercontent.com/6724477/86669558-909f6e80-bfa8-11ea-9e22-5f6441c1d6e0.png)

- Updated Popup / Alternate Clean / Manual Cleaning
![Screenshot_20200707_121826](https://user-images.githubusercontent.com/6724477/86831728-1de7cf00-c04c-11ea-8158-8e41f213576d.png)

- Updated Context Menu
![Screenshot_20200707_121752](https://user-images.githubusercontent.com/6724477/86831764-27713700-c04c-11ea-8558-3152ed16e8f6.png)

kennethtran93 committed:
Added manual cleanup of site data for additional site data types cleanable by hostname via popup and context menus.

Signed-off-by: Kenneth Tran <kennethtran93@users.noreply.github.com>

ke-d commented on src/services/CleanupService.ts:
```diff
@@ -404,110 +406,218 @@ export const clearLocalstorageForThisDomain = async (
+export const clearSiteDataForThisDomain = async (
   state: State,
-  domains: string[],
+  siteData: string,
```
Maybe an enum here would be better?

kennethtran93 replied:
Implemented in commit ac90490

ke-d commented on src/services/CleanupService.ts:
```diff
@@ -404,110 +406,218 @@ export const clearLocalstorageForThisDomain = async (
+
+export const removeSiteData = async (
+  siteData: string,
+  isChrome: boolean,
```
Hmm, maybe not making it specific here would be best (not boolean). Something like a browser enum would be more futureproof for future browsers. Because otherwise, your options here are Chrome and Not Chrome.

Then in the body you can do a switch statement with the different implementations.

kennethtran93 replied:
Implemented in commit ac90490

ke-d commented on src/services/CleanupService.ts:
```diff
@@ -404,110 +406,218 @@ export const clearLocalstorageForThisDomain = async (
+    );
+  }
+  if (
+    getSetting(state, 'indexedDBCleanup') &&
```
Don't have to do it in this PR but just looking at this, I think `indexedDBCleanup` and others for getSetting should be part of an enum string. Otherwise, it would be prone to misspellings and refactoring nightmares (like changing the name).

Just something to note for the future.

ke-d commented on src/services/CleanupService.ts:
```diff
@@ -754,42 +874,44 @@ export const cleanCookiesOperation = async (
+      state,
+      isSafeToCleanObjects,
+    );
+    if (storeResults.cache) {
```
Probably be best to make a helper function here so you don't have to declare `setOfDeletedDomainX` above and repeat the logic here many times.

Like the input would be `storeResults.cache` =>`setOfDeletedDomainCaches`

Another thing is that you can do `new Set(array)`

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set/Set

kennethtran93 replied:
So what you're thinking is to create an array of objects returned from `otherBrowsingDataCleanup`, then outside the container loops, parse and combine the domains from each site data type from all containers and use Set once to remove dups?

kennethtran93 replied:
Implemented in commit ac90490

ke-d commented on src/services/Libs.ts:
```diff
@@ -344,12 +344,15 @@ export const parseCookieStoreId = (
-export const prepareCleanupDomains = (domain: string): string[] => {
+export const prepareCleanupDomains = (
+  domain: string,
+  isChrome: boolean,
```
Same situation as before. Browser enum and switch would be more extendable.

kennethtran93 replied:
Implemented in commit ac90490

ke-d commented on src/services/ContextMenuEvents.ts:
```diff
@@ -447,6 +493,78 @@ export default class ContextMenuEvents extends StoreUser {
+        debug,
+      );
+      switch (siteData) {
+        case 'All':
```
String Enum suggestion here.

kennethtran93 replied:
Implemented in commit ac90490

ke-d commented on src/services/CleanupService.ts:
```diff
@@ -404,110 +406,218 @@ export const clearLocalstorageForThisDomain = async (
+export const cleanSiteData = async (
+  siteData: string,
+  cleanReasonObjects: CleanReasonObject[],
+  isChrome: boolean,
```
Ditto as before.

kennethtran93 replied:
Implemented in commit ac90490

kennethtran93 committed:
Revisions made from PR review comments, and related enhancements.

- SiteData is now an enum type.
- localstorageCleanup setting is now localStorageCleanup to be consistent with other site data types.
- localization related to above has been renamed accordingly.

Signed-off-by: Kenneth Tran <kennethtran93@users.noreply.github.com>

kennethtran93 committed:
Missed committing one additional change for tests to work.

Signed-off-by: Kenneth Tran <kennethtran93@users.noreply.github.com>

kennethtran93 committed:
Missed committing one additional change for tests to work...again.

Signed-off-by: Kenneth Tran <kennethtran93@users.noreply.github.com>

kennethtran93 committed:
Localization name update to reflect more of what that is used for.

Signed-off-by: Kenneth Tran <kennethtran93@users.noreply.github.com>

kennethtran93 committed:
Activity Log Entry Enhancements, plus notification for Site Data Cleanup triggers.

- Activity Log entries now show which site data cleanup was triggered.
- Notifications for site data will appear as a separate notification from cookies.
- Undo button will not show if only site data cleanups were triggered.
- This also adds a button to remove individual log entries.

Signed-off-by: Kenneth Tran <kennethtran93@users.noreply.github.com>

kennethtran93 committed:
ESLint and Auto-Formatting Adjustments on save.

Signed-off-by: Kenneth Tran <kennethtran93@users.noreply.github.com>

kennethtran93 committed:
Fix internal setting check on delayBeforeClean

Signed-off-by: Kenneth Tran <kennethtran93@users.noreply.github.com>

kennethtran93 committed:
Fix cookie restoring logic.  Before this, a dot would be prepended in the domain if it wasn't there before.

Signed-off-by: Kenneth Tran <kennethtran93@users.noreply.github.com>

kennethtran93 committed:
Show Container ID and name in Expression Table and Activity Log.

- This also adds localization for 'Default' for No Containers.

Signed-off-by: Kenneth Tran <kennethtran93@users.noreply.github.com>

kennethtran93 committed:
Default Expression Option Enhancements due to additional site data options.

- Downgrade between 3.4.0 and 3.5.0 should be possible afterwards.
- Deprecation of two settings in favor of using defaults from expression table.
- Core Settings page updated for this.
- Adds a button in Expression Table Settings Page.
- Some minor CSS-related changes.

Signed-off-by: Kenneth Tran <kennethtran93@users.noreply.github.com>

kennethtran93 committed:
Remove locally caught exceptions.

Signed-off-by: Kenneth Tran <kennethtran93@users.noreply.github.com>

kennethtran93 committed:
Convert console line into cadLog.

Signed-off-by: Kenneth Tran <kennethtran93@users.noreply.github.com>

kennethtran93 committed:
Always remove CAD-inserted Browsing Data Cookie to trigger site data removals.

Signed-off-by: Kenneth Tran <kennethtran93@users.noreply.github.com>

kennethtran93 commented:
New Changes committed as of July 16th:

    New Default Expression Options are now moved to an actual expression entry itself. See related documentation.
    (Automatic) Site Data Cleanup triggers are now reflected in Activity Log Entries.
    Notifications for Site Data Cleanup - manual and automatic.
    Undo button will not show for entries with no cookies to restore.
    Remove individual log entries.
    Migration of settings between 3.4.0 and 3.5.0 - should work if for any reason a downgrade is required, minus the added items.

kennethtran93 committed:
Expression List Enhancements
- Ability to remove all expressions from a single list.
- Now shows lists without a matching Firefox container ID.
- Prettier formats

Signed-off-by: Kenneth Tran <kennethtran93@users.noreply.github.com>

kennethtran93 committed:
Filter out 'default' from unknown lists.  Use default container names during default expression creation

Signed-off-by: Kenneth Tran <kennethtran93@users.noreply.github.com>

kennethtran93 commented:
Commit eb77f38 and 994ce6e adds the following Expression List Enhancements  (only if Firefox Contextual Identities/Containers enabled):

- Ability to remove all expressions from a single selected list.
- Now shows lists whose id could not be found in the Firefox container list.

sdht0 commented:
/get-build

ke-d commented on src/services/CleanupService.ts:
```diff
@@ -443,15 +467,25 @@ export const clearSiteDataForThisDomain = async (
-  const listName = isChrome ? 'origins' : 'hostnames';
+  const listName = ((b: string) => {
+    switch (b) {
+      case 'Chrome':
```
Optional for this PR. This can be an enum.

kennethtran93 replied:
Sure.  As long as we don't modify the return string from browserDetect.js in global_files

kennethtran93 replied:
Resolved through 17b7bff

sdht0 commented:
Thanks a lot for these changes!

I tested the new cleanup feature on one website and I'm not sure if CAD is not working or it's still a limitation of the API. Steps:
- Set "Enable Automatic Cleaning" and all "Other Browsing Data Cleanup Options" to true in CAD settings.
- Go to Firefox preferences -> Privacy & Security -> Manage Data... -> Search for "fandom" -> Remove selected -> Save changes.
- Refresh and open "Manage Data..." again to verify that no data for "fandom" is present.
- Open [https://onepiece.fandom.com/wiki/Toki_Toki_no_Mi](https://onepiece.fandom.com/wiki/Toki_Toki_no_Mi). This sets cookies and other storage.
- Refresh and open "Manage Data..." to verify that cookie and storage data is now present for "fandom".
- Close the website. CAD will automatically delete all data.
- Refresh and open "Manage Data...". It can be seen that all cookies were deleted successfully for "fandom", but some storage is still left behind. I am not sure how to check what kind of storage this is.

Some other feedback and suggestions:
* I'm seeing a lot of `Some site data, including 0 Cookie(s), removed from: (Private)` in the cleanup log with no detailed content. Maybe some more context/details can be added as to what was cleaned?
* In "List of Expressions" settings page, instead of putting the container id in all the tabs, maybe display it above the expression table only for the currently selected tab? I feel the tabs are now quite cluttered, esp when there are a lot of containers.
* Minor: instead of "attempted to cleanup...", maybe say "invoked cleanup of..." or something equivalent that is more definitive?

ke-d commented on src/services/CleanupService.ts:
```diff
@@ -443,15 +467,25 @@ export const clearSiteDataForThisDomain = async (
+        return 'hostnames';
+    }
+  })(browserDetect);
+  const sd = `${siteData[0].toLowerCase()}${siteData.slice(1)}`;
```
I think this is common enough to be its own mini function, for the purposes of documentation and easier maintenance in the future.

kennethtran93 replied:
Resolved through 17b7bff

ke-d commented on src/services/Libs.ts:
```diff
@@ -176,6 +184,41 @@ export const getHostname = (urlToGetHostName: string | undefined): string => {
+  storeId: string,
+  listType: ListType,
+): Expression => {
+  const getExpression = (list: string): Expression | undefined => {
```
I haven't done TS in a while, but I think that you can do `Expression?` rather than `Expression | undefined`. If this is wrong, feel free to ignore this comment.

kennethtran93 replied:
What you're referencing is a function return type, so that doesn't work apparently.

kennethtran93 commented:
@sdht0 Thank you very much for assisting with the test builds!  I've left some notes below for you:

>   * Refresh and open "Manage Data...". It can be seen that all cookies were deleted successfully for "fandom", but some storage is still left behind. I am not sure how to check what kind of storage this is.

At this point, the only way I know of at this point, is to have that tab open, use the `Clean, include open tabs` action (via popup clean dropdown or right-click),  First, press `F12` or goto to either the `Tools` menu or the firefox hamburger menu near the top right, then click on `Web Developer` followed by `Storage Inspector`.  Here you should see the list of Cache, Cookies, LocalStorage, IndexedDB, and Session Storage.  Service Workers are under the 'Applications' tab of the same devtools panel.


> Some other feedback and suggestions:
> 
>   * I'm seeing a lot of `Some site data, including 0 Cookie(s), removed from: (Private)` in the cleanup log with no detailed content. Maybe some more context/details can be added as to what was cleaned?

If you're using Private Browsing, that is most likely the cause.  For privacy reasons, we do not output any urls that came from a Private Browsing container.  Though if that is showing up then I think I need to adjust the urls from the site data domains to not include private browsing results.

> 
>   * In "List of Expressions" settings page, instead of putting the container id in all the tabs, maybe display it above the expression table only for the currently selected tab? I feel the tabs are now quite cluttered, esp when there are a lot of containers. 

Unfortunately, due to the fact that we can store outdated container id that doesn't have an actual corresponding Firefox container (thus no name), it won't be easy.  This PR contains the recently committed feature of showing the unresolved container IDs for end-users to remove themselves, especially when exporting and importing from different profiles and/or resetting the container preferences in Firefox.

> Maybe also put "Remove All Expressions from Selected List" in the previous line?

This button only appears in Firefox when containers/contextual Identities is enabled, which falls in line with the container tabs shown as well.  There is already another button to remove all expressions which is used for all lists (if containers is enabled) for the default list if containers disabled or in Chrome.

>   * Minor: instead of "attempted to cleanup...", maybe say "invoked cleanup of..." or something equivalent that is more definitive?

Sounds like a plan.

sdht0 commented:
> Here you should see the list
It looks like both Session storage and Cache storage are not being cleaned up. Not sure about the former but at least the later should've been cleaned, no?

> If you're using Private Browsing, that is most likely the cause.
I am not using it at all. In fact, I've CAD disabled in private browsing mode as it will all get cleaned up anyways. So these logs are ore mysterious in my case.

> Unfortunately, due to the fact that we can store outdated container id that doesn't have an actual corresponding Firefox container (thus no name), it won't be easy.
How about using the names when available and using the id when missing? My goal was to reduce the display clutter somehow, but am not too attached to the solutions I proposed.

> This button only appears in Firefox when containers/contextual Identities is enabled
Yeah, I was suggesting another cosmetic change to reduce the use of vertical space. Moving the button to the previous line will help avoid the white space waste to the right of the buttons. The button logically belongs to the previous line. Of course, in  other browsers it won't be displayed at all.

kennethtran93 commented:
> It looks like both Session storage and Cache storage are not being cleaned up. Not sure about the former but at least the later should've been cleaned, no?

From the domains shown in the Cache section, what remains after cleanup?  The actual domain or third-party domains?  At this point because we don't view network requests, we don't (or rather can't) set any cleanup cookies for third-party domains.  This is already a known issue since somebody else reported that third-party localstorage doesn't get cleared if no cookies from third-party is set.  As for the former, I'm not sure either as there's no specific option for that (only localStorage).

> I am not using it at all. In fact, I've CAD disabled in private browsing mode as it will all get cleaned up anyways. So these logs are more mysterious in my case.
If you expand the summary, do you see any specific URLs from any specific site data that was cleaned up?

> How about using the names when available and using the id when missing? My goal was to reduce the display clutter somehow, but am not too attached to the solutions I proposed.

While this may work, people can set duplicate names thus only the id is unique.  I'll try and move the ID somewhere else, but it will still be shown regardless.  This is to help users that want to move expressions as until I implement that feature, users need to manually export, change the file itself, then re-import it.  The Import only relies on the container ID.

> Yeah, I was suggesting another cosmetic change to reduce the use of vertical space. Moving the button to the previous line will help avoid the white space waste to the right of the buttons. The button logically belongs to the previous line. Of course, in other browsers it won't be displayed at all.

I'll give that a shot.

sdht0 commented:
> The actual domain or third-party domains

The actual domain. It should be pretty easy to reproduce at your end as per the steps I outlined above and help you debug if the issue is on CAD's end.

>  If you expand the summary, do you see any specific URLs from any specific site data that was cleaned up?

The details are empty (the first one is expanded):
![image](https://user-images.githubusercontent.com/867424/87894002-7be2c200-ca0f-11ea-99e0-d8c615c5871c.png)

>  move the ID somewhere else, but it will still be shown regardless

Yes this is what I meant. Instead of showing it for all the tabs at once, only show it for the active tab (just above the expression table is one option). Again, this was just a suggestion and I can live with the current implementation if it will mean shipping these changes faster.

kennethtran93 commented:
> > The actual domain or third-party domains
> 
> The actual domain. It should be pretty easy to reproduce at your end as per the steps I outlined above and help you debug if the issue is on CAD's end.

Unfortunately there's not much we can debug about these.  With the exception of cookies, no other site data can be queried/enumerated through the webextension API.  All we do is pass in the hostname to be cleaned, and the type of site data to be cleaned.  (API in question is this:  https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/browsingData/remove)

> 
> > If you expand the summary, do you see any specific URLs from any specific site data that was cleaned up?
> 
> The details are empty (the first one is expanded):
> ![image](https://user-images.githubusercontent.com/867424/87894002-7be2c200-ca0f-11ea-99e0-d8c615c5871c.png)


I'm adding a check to remove private containers from the siteData domains, thus wouldn't trigger the flag that adds it to the activity log even if there wasn't any cookies deleted.  I think that's the cause.

kennethtran93 committed:
Addressing changes requested through PR.

- browserDetect return is now an enum type.
- mini function for making siteDataType first letter lowercase (browser property key cleanup).
- notifications for manual cleaning of site data now in CleanupService instead of Popup for consistency with cookies.
- Site Data Cleanup message now more 'definitive'
- Container ID in List of Expressions moved above the tabs, so only names are left in tabs.
- Moved Remove expressions by container to same area as remove all expressions and create default expressions.

Signed-off-by: Kenneth Tran <kennethtran93@users.noreply.github.com>

kennethtran93 commented:
@sdht0 Incorporated some of the changes you mentioned.  Feel free to test it out again.  There's an active bug in regards of `use_site` with `firstparty.isolate`, which I need to work on. It will be on a new branch, that is based off of this however.

ke-d commented on extension/global_files/browserDetect.js:
```diff
@@ -6,67 +6,90 @@
  *
  * @returns {string}
  */
-var browserDetect = function() {
```
Maybe something to look into the future, but maybe there's a npm module that can be imported to do this functionality? However, no action needed in this PR. 

My original reasoning for storing it into redux cache is so that the background and options scripts don't have to recompute the browser detect. But now I'm thinking of ways of reducing complexity for the future.

ke-d reviewed (dismissed):
Looks good to me for the new changes.

sdht0 commented:
Thanks for the changes! Looks good to me.
One issue that I found is that notifications (3 notifications on cleanup for a domain) are being triggered even when the option is OFF in my settings.

kennethtran93 commented:
If you're doing manual cleanup then notification will always show, per an older request. There's a new issue that wants that reverted so in the future I'll need to add another option for that.

sdht0 commented:
Sorry I should have clarified. I meant that automatic cleanup is showing notifications and looks like a regression. Changing the option in settings to ON or OFF does not seem to have an effect, although it looks like when the settings is ON, there are a lot more notifications that are being shown.

kennethtran93 commented:
I'll take a look soon. If the notification has a red icon then it is usually an error thrown that is always displayed. Otherwise I'll need to dig deeper into it and follow the logic.

sdht0 commented:
Yes, looks like the notifications for normal cleanups, with individual notifications for each type. It seems to be a regression in the refactoring done in the last commit. Can you not reproduce this on your end?

kennethtran93 committed:
Adds option to toggle display of manual cleaning of single domains.  This also fixes a notification regression from previous commit.

Signed-off-by: Kenneth Tran <kennethtran93@users.noreply.github.com>

kennethtran93 commented:
@sdht0 Whoops.  Should be fixed now.  I always have notifications turned onI've also added an option to toggle manual notifications display per other requests.  I've tested the relevant notification options and I believe it should be working properly now.

kennethtran93 committed:
Update Release Notes

Signed-off-by: Kenneth Tran <kennethtran93@users.noreply.github.com>

kennethtran93 commented:
Unless there's more regressions, I promise that this will be merged upon approval this time.

ke-d commented on src/services/CleanupService.ts:
```diff
@@ -361,26 +361,33 @@ export const clearCookiesForThisDomain = async (
       if (r) cookieDeletedCount += 1;
     }
-    showNotification({
+    showNotification(
```
Hmm, it might be a good idea to decouple some of the notification logic away from Cleanup service. It can be a wrapper function around `showNotification` that takes in something that Cleanup service would return (maybe CleanReasonObject with some refactoring to include the other side data hostnames). Based on whats in that object, you can output a different types of notifications. Though this would require some refactoring that I don't expect in this PR.

sdht0 commented:
Thanks a lot for all the work @kennethtran93!

fabianski7 commented:
this PR allows [FileSystem](https://developer.mozilla.org/en-US/docs/Web/API/FileSystem) to be cleaned too?

kennethtran93 commented:
> this PR allows [FileSystem](https://developer.mozilla.org/en-US/docs/Web/API/FileSystem) to be cleaned too?

Not yet, per https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/browsingData/DataTypeSet, at least in Firefox.  Seems like only Chrome/Edge Chromium has support for it at this time.

Perhaps in a future release will it be implemented.

kennethtran93 commented:
/get-build
